cpp/lib/Authenticator.cpp

```diff
@@ -337,12 +353,200 @@ namespace Client
+          AUTH_THROW(err->msg);
+      }
+
+      void* curl_desc;
+      CURL* curl;
+      curl_desc = get_curl_desc_from_pool(token_url_str, m_connection->proxy, m_connection->no_proxy);
+      curl = get_curl_from_desc(curl_desc);
+
+      // When renewTimeout < 0, throws Exception
+      // for each retry attempt so we can renew the one time token
+      int64 retry_timeout = get_login_timeout(m_connection);
+      int64 elapsed_time = 0;
+      int8* retried_count = &m_connection->retry_count;
+
+      // 3. get one time token from okta
+      cJSON* body = snowflake_cJSON_CreateObject();
+      snowflake_cJSON_AddStringToObject(body, "username", m_connection->user);
+      snowflake_cJSON_AddStringToObject(body, "password", m_connection->password);
+
+      char* s_body = snowflake_cJSON_Print(body);
+
+      // headers for step 4
+      // add header for accept format
+      SF_HEADER* header = sf_header_create();
+      header->use_application_json_accept_type = SF_BOOLEAN_TRUE;
+      if (!create_header(m_connection, header, &m_connection->error)) {
+          SET_SNOWFLAKE_ERROR(err, SF_STATUS_ERROR_GENERAL, "Failed to created the header for the okta authentication", SF_SQLSTATE_GENERAL_ERROR);
+          AUTH_THROW(err);
+      }
+      cJSON* resp = NULL;
+
+      if (!curl_post_call(m_connection, curl, token_url_str, header, s_body, &resp,
+          &m_connection->error, 0, get_login_retry_count(m_connection), retry_timeout,
+          &elapsed_time, retried_count, 0, false))
+      {
+          CXX_LOG_WARN("sf", "AuthenticatorOKTA", "getSamlResponseUsingOkta",
+              "Fail to get SAML response, response body=%s",
+              snowflake_cJSON_Print(resp));
+          SET_SNOWFLAKE_ERROR(err, SF_STATUS_ERROR_BAD_REQUEST, "OktaConnectionFailed", SF_SQLSTATE_GENERAL_ERROR);
+          AUTH_THROW(err);
+      }
+
+      char* one_time_token = snowflake_cJSON_HasObjectItem(resp, "sessionToken") ?
+          snowflake_cJSON_GetStringValue(snowflake_cJSON_GetObjectItem(resp, "sessionToken")) :
+          snowflake_cJSON_GetStringValue(snowflake_cJSON_GetObjectItem(resp, "cookieToken"));
+      if (elapsed_time >= retry_timeout)
+      {
+          CXX_LOG_WARN("sf", "AuthenticatorOKTA", "getSamlResponseUsingOkta",
+              "Fail to get SAML response, timeout reached: %d, elapsed time: %d",
+              retried_count, elapsed_time);
+          SET_SNOWFLAKE_ERROR(err, SF_STATUS_ERROR_REQUEST_TIMEOUT, "OktaConnectionFailed: timeout reached", SF_SQLSTATE_GENERAL_ERROR);
+          AUTH_THROW(err);
+      }
+
+      // 4. get SAML response
+      SFURL sso_url = SFURL::parse(sso_url_str);
+      sso_url.addQueryParam("onetimetoken", one_time_token);
+      resp = NULL;
+      if (!curl_get_call(m_connection, curl, (char*)sso_url.toString().c_str(), NULL, &resp, &m_connection->error))
```

Comment:
We need renew one time token for each retry on get request. Check ODBC PR for details. https://github.com/snowflakedb/snowflake-odbc/pull/509

Reply:
I applied it to the implementation.